src/types/resource/stream.rs

```diff
@@ -92,6 +96,82 @@ impl Stream {
                 behavior_hints: Default::default(),
             })
     }
+    pub fn download_url(&self) -> Option<String> {
+        match &self.source {
+            StreamSource::Url { url } if url.scheme() == "magnet" => {
+                self.magnet_url().map(|magnet_url| magnet_url.to_string())
+            }
+            StreamSource::Url { url } => Some(url.to_string()),
+            StreamSource::Torrent { .. } => {
+                self.magnet_url().map(|magnet_url| magnet_url.to_string())
+            }
+            StreamSource::YouTube { .. } => self.youtube_url(),
+            StreamSource::External { external_url, .. } => {
+                external_url.as_ref().map(|url| url.to_string())
+            }
+            StreamSource::PlayerFrame { player_frame_url } => Some(player_frame_url.to_string()),
+        }
+    }
+    pub fn m3u_data_uri(&self, streaming_server_url: Option<&Url>) -> Option<String> {
+        self.streaming_url(streaming_server_url).map(|url| {
+            format!(
+                "data:application/octet-stream;charset=utf-8;base64,{}",
+                base64::encode(format!("#EXTM3U\n#EXTINF:0\n{}", url))
+            )
+        })
+    }
+    pub fn streaming_url(&self, streaming_server_url: Option<&Url>) -> Option<String> {
+        match (&self.source, streaming_server_url) {
+            (StreamSource::Url { url }, _) if url.scheme() != "magnet" => Some(url.to_string()),
+            (
+                StreamSource::Torrent {
+                    info_hash,
+                    file_idx,
+                    announce,
+                },
+                Some(streaming_server_url),
+            ) => {
+                let mut url = streaming_server_url.to_owned();
+                match url.path_segments_mut() {
+                    Ok(mut path) => {
+                        path.push(&hex::encode(info_hash));
+                        if let Some(file_idx) = file_idx {
```

Comment:
@tymmesyde If I'm reading correctly, if `file_index` is `None` then the created url won't contain the file index path and will be just `/infoHash`? This is not a supported path by enginefs as far as I know, since you always need to provide and an index, correct  @jaruba ? 
So if `file_index` is `None` you can't really construct a streaming url, as first you need to do a POST request to enginefs to resolve the largest file and get it's index, which you can then use to construct such GET url.